modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,237 @@
+	 * @return array<int, ILO_URL_Metric[]> URL metrics grouped by breakpoint. The array keys are the minimum widths for a viewport to lie within
+	 *                                      the breakpoint. The returned array is always one larger than the provided array of breakpoints, since
+	 *                                      the breakpoints reflect the max inclusive boundaries whereas the return value is the groups of page
+	 *                                      metrics with viewports on either side of the breakpoint boundaries.
+	 */
+	private function group_url_metrics_by_breakpoint( array $url_metrics ): array {
+
+		// Convert breakpoint max widths into viewport minimum widths.
+		$minimum_viewport_widths = array_map(
+			static function ( $breakpoint ) {
+				return $breakpoint + 1;
+			},
+			$this->breakpoints
+		);
+
+		$grouped = array_fill_keys( array_merge( array( 0 ), $minimum_viewport_widths ), array() );
+
+		foreach ( $url_metrics as $url_metric ) {
+			$current_minimum_viewport = 0;
+			foreach ( $minimum_viewport_widths as $viewport_minimum_width ) {
+				if ( $url_metric->get_viewport()['width'] >= $viewport_minimum_width ) {
+					$current_minimum_viewport = $viewport_minimum_width;
+				} else {
+					break;
+				}
+			}
+
+			$grouped[ $current_minimum_viewport ][] = $url_metric;
+		}
+		return $grouped;
+	}
+
+	/**
+	 * Determines whether the URL metrics group for a given viewport is lacking URL metrics.
+	 *
+	 * Either the group does not have enough URL metrics for the desired sample size,
+	 * or some of the URL metrics are stale.
+	 *
+	 * @param int $viewport_width Viewport width.
+	 * @return bool Whether group is lacking.
+	 */
+	public function is_group_lacking( int $viewport_width ): bool {
+		$last_was_lacking = false;
+		foreach ( $this->get_group_statuses() as $status ) {
+			if ( $viewport_width >= $status->get_minimum_viewport_width() ) {
+				$last_was_lacking = $status->is_lacking();
+			} else {
+				break;
+			}
+		}
+		return $last_was_lacking;
+	}
+
+	/**
+	 * Gets the statuses of the URL metrics groups.
+	 *
+	 * A group is lacking URL metrics if it does not have the desired sample size or
+	 * if some of the URL metrics are stale.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @return ILO_URL_Metrics_Group_Status[] URL metrics group statuses.
+	 */
+	public function get_group_statuses(): array {
+		$current_time = microtime( true );
+
+		$statuses = array();
+		foreach ( $this->groups as $minimum_viewport_width => $viewport_url_metrics ) {
+			$is_lacking = false;
+			if ( count( $viewport_url_metrics ) < $this->sample_size ) {
+				$is_lacking = true;
+			} else {
+				foreach ( $viewport_url_metrics as $url_metric ) {
+					if ( $url_metric->get_timestamp() + $this->freshness_ttl < $current_time ) {
+						$is_lacking = true;
+						break;
+					}
+				}
+			}
+			$statuses[] = new ILO_URL_Metrics_Group_Status( $minimum_viewport_width, $is_lacking );
+		}
+
+		return $statuses;
+	}
```

Comment:
See my above comments:
* Why does a group status not also include its upper (maximum) viewport width?
* I think rather than a more specific "URL metrics group status" class, we should have a "URL metrics group" class.
    * We already have "URL metric" and "grouped URL metrics", so IMO "URL metrics group" is the missing piece between the two. A class for their status would be more specific, and IMO is unnecessary as it could be implemented as part of the group class.
    * We could then create those instances as part of the `group_url_metrics_by_breakpoint()` method, and the `get_groups()` method would be the logical place to return those objects from - or maybe another one, if we wanted to keep a method in place to return _just_ the minimum viewport widths.

Based on that, I think this method could be replaced with something simpler. That said, we should probably first figure out the problem of not considering maximum viewport widths here - maybe I'm missing something but I don't see why those aren't being considered.

Reply:
+1. Coincidentally, after I put that `ILO_URL_Metrics_Group_Status` class together on Friday, over the weekend I was ruminating on it a bit  more and also came to the conclusion that it would make more sense to have an `ILO_URL_Metrics_Group` class which exposes the status as an aspect. I didn't get a chance to follow up until now.